server/src/main/kotlin/com/xebia/functional/xef/server/ai/providers/MLflowApiProvider.kt

```diff
@@ -0,0 +1,85 @@
+package com.xebia.functional.xef.server.ai.providers
+
+import com.xebia.functional.xef.server.ai.copyFrom
+import com.xebia.functional.xef.server.ai.providers.mlflow.PathProvider
+import com.xebia.functional.xef.server.ai.providers.mlflow.mlflowPathProvider
+import com.xebia.functional.xef.server.models.exceptions.XefExceptions
+import com.xebia.functional.xef.server.models.mlflow.*
+import io.ktor.client.*
+import io.ktor.client.request.*
+import io.ktor.client.statement.*
+import io.ktor.http.*
+import io.ktor.server.application.*
+import io.ktor.server.auth.*
+import io.ktor.server.response.*
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.*
+import org.slf4j.Logger
+
+@OptIn(ExperimentalSerializationApi::class)
+class MLflowApiProvider(
+    private val gatewayUrl: String,
+    private val pathProvider: PathProvider,
+    private val client: HttpClient
+): ApiProvider {
+
+    override suspend fun chatRequest(call: ApplicationCall, requestBody: ByteArray) {
+        val stringRequestBody = requestBody.toString(Charsets.UTF_8)
+        val requestData = Json.decodeFromString<JsonObject>(stringRequestBody)
+        val isStream = requestData["stream"]?.jsonPrimitive?.boolean ?: false
+        if (isStream) {
+            call.respond(HttpStatusCode.NotImplemented, "Stream not supported")
+        } else {
+            val model = requestData["model"]!!.jsonPrimitive.content
+            val principal = call.principal<UserIdPrincipal>() ?: throw XefExceptions.AuthorizationException("")
+            val path = pathProvider.chatPath(model, principal)
+            val newData = requestData.filterKeys { !filterOutFields.contains(it) }.mapKeys {mappedFields.getOrDefault(it.key, it.key) }
```

Comment:
I do not know exactly about the API specification and how it relates to the OAI spec.
Though, does it make more sense to specify _what fields are needed_ rather than which are not?

Reply:
I don't know. To give more context, mlflow gateway throws an error when you send some parameters that the API is not expecting (like the `model`). On the other hand, having a list of the needed fields will require us to change the list if new fields are added. With the current implementation, if a new field is added, the clients can start sending them immediately. Do you think it makes sense?